scapy/layers/ldap.py

```diff
@@ -0,0 +1,444 @@
+        ASN1F_SEQUENCE_OF("attributes",
+                          LDAP_SearchResponseEntryAttribute(),
+                          LDAP_SearchResponseEntryAttribute)
+    )
+
+
+class LDAP_SearchResponseResultCode(ASN1_Packet):
+    ASN1_codec = ASN1_Codecs.BER
+    ASN1_root = LDAPResult
+
+
+class LDAP_AbandonRequest(ASN1_Packet):
+    ASN1_codec = ASN1_Codecs.BER
+    ASN1_root = ASN1F_INTEGER("messageID", 0)
+
+
+# LDAP v3
+
+
+class LDAP_Control(ASN1_Packet):
+    ASN1_codec = ASN1_Codecs.BER
+    ASN1_root = ASN1F_SEQUENCE(
+        LDAPOID("controlType", ""),
+        ASN1F_optional(
+            ASN1F_BOOLEAN("criticality", False),
+        ),
+        ASN1F_optional(
+            ASN1F_STRING("controlValue", "")
+        ),
+    )
+
+
+# LDAP
+
+
+class LDAP(ASN1_Packet):
+    ASN1_codec = ASN1_Codecs.BER
+    ASN1_root = ASN1F_SEQUENCE(
+        ASN1F_INTEGER("messageID", 0),
+        ASN1F_CHOICE("protocolOp", LDAP_SearchRequest(),
+                     ASN1F_PACKET("bindRequest",
+                                  LDAP_BindRequest(),
+                                  LDAP_BindRequest,
+                                  implicit_tag=0x60),
+                     ASN1F_PACKET("bindResponse",
+                                  LDAP_BindResponse(),
+                                  LDAP_BindResponse,
+                                  implicit_tag=0x61),
+                     ASN1F_PACKET("unbindRequest",
+                                  LDAP_UnbindRequest(),
+                                  LDAP_UnbindRequest,
+                                  implicit_tag=0x42),
+                     ASN1F_PACKET("searchRequest",
+                                  LDAP_SearchRequest(),
+                                  LDAP_SearchRequest,
+                                  implicit_tag=0x63),
+                     ASN1F_PACKET("searchResponse",
+                                  LDAP_SearchResponseEntry(),
+                                  LDAP_SearchResponseEntry,
+                                  implicit_tag=0x64),
+                     ASN1F_PACKET("searchResponse",
+                                  LDAP_SearchResponseResultCode(),
+                                  LDAP_SearchResponseResultCode,
+                                  implicit_tag=0x65),
+                     ASN1F_PACKET("abandonRequest",
+                                  LDAP_AbandonRequest(),
+                                  LDAP_AbandonRequest,
+                                  implicit_tag=0x70)
+                     ),
+        # LDAP v3 only
+        ASN1F_optional(
+            ASN1F_SEQUENCE_OF("Controls", [], LDAP_Control,
+                              implicit_tag=0x0)
+        )
+    )
+
+    def mysummary(self):
+        return (self.protocolOp.__class__.__name__.replace("_", " "), [LDAP])
+
+
+bind_layers(LDAP, LDAP)
+
+bind_bottom_up(TCP, LDAP, dport=389)
+bind_bottom_up(TCP, LDAP, sport=389)
+bind_layers(TCP, LDAP, sport=389, dport=389)
```

Comment:
I'm not sure why we should default to setting both source and destination ports to 389?

Reply:
Scapy has this weird quirk that because `UDP` has port 53 in its default value, not setting both ports create packets that are wrongly forced-dissected as DNS. See the demo below

```python
>>> class A(Packet):
...:     fields_desc=[ByteField("a", 0)]
...: 
>>> bind_layers(UDP, A, dport=389)
>>> (UDP()/A()).show()
###[ UDP ]### 
  sport     = domain
  dport     = ldap
  len       = None
  chksum    = None
###[ A ]### 
     a         = 0

>>> (UDP()/A()).show2()
INFO: No IP underlayer to compute checksum. Leaving null.
###[ UDP ]### 
  sport     = domain
  dport     = ldap
  len       = 9
  chksum    = 0x0
###[ Raw ]### 
     load      = '\x00'

>>> bind_layers(UDP, A, sport=389)
>>> (UDP()/A()).show()
###[ UDP ]### 
  sport     = ldap
  dport     = domain
  len       = None
  chksum    = None
###[ A ]### 
     a         = 0

>>> (UDP()/A()).show2()
INFO: No IP underlayer to compute checksum. Leaving null.
###[ UDP ]### 
  sport     = ldap
  dport     = domain
  len       = 9
  chksum    = 0x0
###[ Raw ]### 
     load      = '\x00'
```

We can't really change the default port value now without breaking everything, so the work around is for the last binding to set both ports to make sure this doesn't happen.

With TCP, the default port is 80 which by default isn't bound to anything so this is less of an issue, but loading the optional `scapy.layers.http` layer does bind 80 which then comes down to the same issue.

Reply:
We should really have a `EphemeralPort()` value but that's OK for me in the meantime. Thanks for this explanation!